configdialog.cpp

```diff
@@ -518,6 +535,271 @@ void PaletteCustomizeDialog::get_and_save_color(int palette_index) {
 	}
 }
 
+std::wstring PaletteCustomizeDialog::export_palettes() {
+	auto all_palettes = m_palette_repo.get_all_custom_palettes();
+	std::wstring palettes_string = L"";
+
+	for (const auto &palette : all_palettes) {
+		auto name = palette.name;
+		auto *data = palette.data;
+
+		palettes_string.append(palette.name + L": ");
+		std::wstring colors = m_palette_repo.serialize(*data) + L";";
+		// Add extra "fake" colors so that a palette definition is guaranteed to be longer than the maximum allowed name length
+		while (colors.length() < PaletteCustomizeDialog::MaxPaletteNameSize) {
+			colors.append(L"#000000;");
+		}
+		palettes_string.append(colors);
+		palettes_string.append(L"\r\n");
+	}
+	palettes_string.append(L"~");
+
+	return palettes_string;
+}
+
+void PaletteCustomizeDialog::import_palettes(std::wstring *palettes) {
+	using namespace std;
+
+	const static wstring valid_chars = L"qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM0123456789-_!?:";
+	const static wstring valid_hex_chars = L"0123456789abcdefABCDEF";
+	const static size_t colors_amount = max(7, (size_t)ceil(PaletteCustomizeDialog::MaxPaletteNameSize / 8.0));
+
+	size_t palettes_amount = 0;
+	size_t current_index = 0;
+	bool graceful_exit = false;
+	size_t count_invalid = 0;
+
+	const enum class Errors {
+		none = 0,
+		name_too_short = 1 << 0,
+		name_too_long = 1 << 1,
+		name_invalid_characters = 1 << 2,
+		colors_too_few = 1 << 3,
+		colors_incorrect_length = 1 << 4,
+		colors_invalid_characters = 1 << 5,
+	};
+
+	vector<wstring> names;
+	vector<vector<wstring>> colors;
+	vector<uint8_t> is_valid;
+
+	for (; palettes->find_first_of(valid_chars, current_index) != wstring::npos && palettes->find_first_of(L':', current_index) != wstring::npos; palettes_amount++) {
+		graceful_exit = (palettes->find_first_of(L'~', current_index) != wstring::npos);
```

Comment:
Again, I'm not going to comb over this line-by-line, because it hurts me. It just needs to work for now - we will determine that by testing it at runtime, once everything is done.